src/lib/asn1/asn1_obj.h

```diff
@@ -19,17 +19,24 @@ class BER_Decoder;
 class DER_Encoder;
 
 /**
-* ASN.1 Type and Class Tags
+* ASN.1 Class Tags
 */
-enum class ASN1_Tag : uint32_t {
-   UNIVERSAL        = 0x00,
-   APPLICATION      = 0x40,
-   CONTEXT_SPECIFIC = 0x80,
+enum class ASN1_Class : uint32_t {
+    UNIVERSAL        = 0b0000'0000,
+    APPLICATION      = 0b0100'0000,
+    CONTEXT_SPECIFIC = 0b1000'0000,
+    PRIVATE          = 0b1100'0000,
 
-   CONSTRUCTED      = 0x20,
+    CONSTRUCTED      = 0b0010'0000,
+    EXPLICIT_CONTEXT_SPECIFIC = CONSTRUCTED | CONTEXT_SPECIFIC,
 
-   PRIVATE          = CONSTRUCTED | CONTEXT_SPECIFIC,
+    NO_OBJECT        = 0xFF00
```

Comment:
I didn't know how to correctly handle code that used this enum value so both enum classes have this value.
Maybe instead of a sentinel value, use an `std::optional`? Sadly, I am not very familiar with your code base so I can't make any serious proposal here.

Reply:
Using `std::optional` is probably viable. Only a few places care about or have to consider an unset tag (mostly in the decoder)